### PR TITLE
Define Pydantic schema on `Quantity`

### PR DESCRIPTION
**Pull request by mattwthompson**
## Description
This is a possible solution for #177
 and https://github.com/openforcefield/openff-interchange/issues/1346 and arguably a better design from the ground-up than what we had in `openff-models`

## Todos
  - [x] Handle different inputs
    - [x] `str`
    - [x] `Quantity`
    - [x] `list` or similar iterable
    - [x] `dict`
    - [x] `numpy.ndarray`
  - [x] Test different serialization pathways
    - [x] Python
    - [x] JSON
  - [x] Test different de-serialization pathways
    - [x] Python
    - [x] JSON
  - [ ] Test unit/dimensionality edge cases?
  - [x] Make Pydantic optional?
  - [x] Test against other OpenFF libraries: https://github.com/openforcefield/status/pull/147
## Status
- [ ] Ready to go

### CODE REVIEW

**Review by j-wags — COMMENTED** [review]
I spoke with the OpenFE leads and concluded that we should move forward with this approach. Please let me know once this is ready for review!

File: openff/units/units.py [hunk 1]
```diff
@@ -32,7 +45,120 @@ class Unit(pint.UnitRegistry.Unit):
...
+                magnitude = v.magnitude
+
+                # storing numpy arrays natively works fine in memory in Python,
+                # but must be list-ified when serializing to JSON.
+                if isinstance(magnitude, numpy.ndarray):
+                    magnitude = v.magnitude.tolist()
+
+                # TODO: I think this is necessary for handling unit-wrapped arrays, but it is
+                #       not so performant. Scalar quantities can be directly serialized to much
+                #       shorter strings
```
Comment (j-wags) [inline openff/units/units.py:69]: (not blocking) If performance is a concern, maybe add a test defining/serializing/deserializing a big array quantity and test performance with a timeout (even if the current performance is acceptable, this could prevent other changes in the future from exacerbating performance issues)

File: openff/units/units.py [hunk 2]
```diff
@@ -32,7 +45,120 @@ class Unit(pint.UnitRegistry.Unit):
...
+            if isinstance(v, Quantity):
+                return v
+            elif isinstance(v, str):
+                # TODO: A significant wart is that we have to try to guess whether the string is a
+                #       JSON-serialized quantity or a simple string representation of a quantity.
+                #       For example: input of "0.9 nanometer" can be passed directly to the
+                #       Quantity constructor, but "{"magnitude": 0.9, "units": "nanometer"}" cannot
+                #       as it needs to be unwrapped. A better solution would require a better way
+                #       of serializing unit-wrapped arrays to JSON
+                if "{" in v:
```
Comment (j-wags) [inline openff/units/units.py:96]: (not blocking) 👍 I was certain I was about to see an `eval` statement after this, nice work not needing that.

File: openff/units/units.py [hunk 3]
```diff
@@ -45,7 +171,7 @@ def _dask_finalize(results, func, args, units):
 
 
 @requires_package("openmm")
-def _to_openmm(self) -> "openmm.unit.Quantity":
+def _to_openmm(self) -> openmm.unit.Quantity:
```
Comment (j-wags) [inline openff/units/units.py:174]: (not blocking) You'd asked me during synchronous review to flag this.

**Review by j-wags — APPROVED** [review]
Thanks @mattwthompson - This looks good to me. A few comments but none blocking. Thanks for reviewing this with me live today.
 
- (not blocking) If the previous version supported list of lists of uneven length (eg `[[1,2],[3]]`) it would be good to either ensure that works after these changes, or test that defining/serializing/deserializing such a value raises an error.

File: openff/units/units.py (hunk 3 above)
Comment (mattwthompson) [reply openff/units/units.py:174]: I was concerned this would break things if OpenMM isn't installed, but that's not the case because I turned on deferred annotations (`from __future__ import annotations`)

```shell
[openff-units] pixi run -e py312pint24 python         define-pydantic-schema  ✱
Python 3.12.13 | packaged by conda-forge | (main, Mar  5 2026, 17:06:14) [Clang 19.1.7 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import openmm
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'openmm'
>>> from openff.units import *
>>> Quantity("1 picosecond")
<Quantity(1, 'picosecond')>

File: openff/units/units.py (hunk 1 above)
Comment (mattwthompson) [reply openff/units/units.py:69]: Valid concern but deferring this until it's diagnosed as an issue
